onnxruntime/core/providers/cpu/controlflow/if.cc

```diff
@@ -193,27 +192,34 @@ Status IfImpl::Execute() {
     }
   }
   std::vector<MLValue> fetches;
+  std::unordered_map<size_t, IExecutor::CustomAllocator> fetch_allocators;
   fetches.reserve(num_outputs_);
 
   for (int i = 0; i < num_outputs_; ++i) {
     fetches.push_back(outputs_[i].second);
-  }
-
-  status = utils::ExecuteGraph(session_state_, feeds, subgraph_output_names_, fetches, /*sequential_execution*/ true,
-                               context_.GetTerminateFlag(), context_.Logger());
-  ORT_RETURN_IF_ERROR(status);
 
-  for (int i = 0; i < num_outputs_; ++i) {
-    // TODO: Task 1913: Improve handling of If outputs to avoid copy when the shape is not known
-    //       at subgraph execution time.
-    if (outputs_[i].first == AllocationType::Temporary) {
-      // need to allocate If output and copy MLValue from fetches
-      auto& data = fetches[i].Get<Tensor>();
-      Tensor* output = context_.Output(i, data.Shape());
-      memcpy(output->MutableDataRaw(), data.DataRaw(), data.Size());
+    if (outputs_[i].first == AllocationType::Delayed) {
+      // functor to forward the allocation request from the subgraph to the If node's context so that the
+      // allocation plan for the If node's output is used.
+      fetch_allocators[i] =
```

Comment:
What is fetch_allocators' index supposed to be? The earlier usage seemed to be a mlvalue-index, while here it looks to be different.

Reply:
If it is an input/output number (as used here), I suppose we would be able to use a vector, instead of a map, which would be more efficient.

Reply:
The index matches an the ordering in feeds.

Is there a good way to have an optional std::function/lambda in a vector in C++14? 

---
In reply to: [251689123](https://github.com/Microsoft/onnxruntime/pull/377#discussion_r251689123) [](ancestors = 251689123)